src/linearcombination.jl

```diff
@@ -39,106 +42,82 @@ julia> LinearMap(ones(Int, 3, 3)) + CS + I + rand(3, 3);
-    A_mul_B!(y, A.maps[1], x)
-    l = length(A.maps)
-    if l>1
-        z = similar(y)
-        for n in 2:l
-            A_mul_B!(z, A.maps[n], x)
-            y .+= z
-        end
+for Atype in (AbstractVector, AbstractMatrix)
+    @eval Base.@propagate_inbounds function LinearAlgebra.mul!(y::$Atype, A::LinearCombination, x::$Atype,
+                             α::Number=true, β::Number=false)
+        @boundscheck check_dim_mul(y, A, x)
+        return _lincombmul!(y, A, x, α, β)
     end
-    return y
 end
 
-else # 5-arg mul! is available for matrices
-
-# map types that have an allocation-free 5-arg mul! implementation
-const FreeMap = Union{MatrixMap,UniformScalingMap}
-
-function A_mul_B!(y::AbstractVector, A::LinearCombination{T,As}, x::AbstractVector) where {T, As<:Tuple{Vararg{FreeMap}}}
-    # no size checking, will be done by individual maps
-    A_mul_B!(y, A.maps[1], x)
-    for n in 2:length(A.maps)
-        mul!(y, A.maps[n], x, true, true)
-    end
-    return y
-end
-function A_mul_B!(y::AbstractVector, A::LinearCombination, x::AbstractVector)
-    # no size checking, will be done by individual maps
-    A_mul_B!(y, A.maps[1], x)
-    l = length(A.maps)
-    if l>1
-        z = similar(y)
-        for n in 2:l
-            An = A.maps[n]
-            if An isa FreeMap
-                mul!(y, An, x, true, true)
-            else
-                A_mul_B!(z, A.maps[n], x)
-                y .+= z
-            end
+@inline function _lincombmul!(y, A::LinearCombination{<:Any,FiveArg}, x, α::Number, β::Number)
+    if iszero(α) # trivial cases
+        iszero(β) && (fill!(y, zero(eltype(y))); return y)
+        isone(β) && return y
+        # β != 0, 1
+        rmul!(y, β)
+        return y
+    else
+        mul!(y, first(A.maps), x, α, β)
+        @inbounds for An in Base.tail(A.maps)
+            mul!(y, An, x, α, true)
         end
+        return y
     end
-    return y
 end
 
-function LinearAlgebra.mul!(y::AbstractVector, A::LinearCombination{T,As}, x::AbstractVector, α::Number=true, β::Number=false) where {T, As<:Tuple{Vararg{FreeMap}}}
-    length(y) == size(A, 1) || throw(DimensionMismatch("mul!"))
-    if isone(α)
-        iszero(β) && (A_mul_B!(y, A, x); return y)
-        !isone(β) && rmul!(y, β)
-    elseif iszero(α)
+@inline function _lincombmul!(y, A::LinearCombination{<:Any,ThreeArg}, x, α::Number, β::Number)
+    if iszero(α)
         iszero(β) && (fill!(y, zero(eltype(y))); return y)
         isone(β) && return y
         # β != 0, 1
         rmul!(y, β)
         return y
-    else # α != 0, 1
-        if iszero(β)
-            A_mul_B!(y, A, x)
-            rmul!(y, α)
-            return y
-        elseif !isone(β)
-            rmul!(y, β)
-        end # β-cases
-    end # α-cases
-
-    for An in A.maps
-        mul!(y, An, x, α, true)
+    else
+        mul!(y, first(A.maps), x, α, β)
+        l = length(A.maps)
+        if l>1
+            z = similar(y)
+            @inbounds for n in 2:l
```

Comment:
Above you used `@inbounds for An in Base.tail(A.maps)` which is nice.

Unfortunately, both options lead to a type unstable `An` for a heterogenous list of maps, which might cause some slowdown for small problems (I expect this to be negligible for larger dimensions). I know `map` works on tuples (if they are not too long) by recursion, such that there is no type instability for heterogeneous tuples, but when you don't care about the result, there is `foreach` which unfortunately also uses a plain loop.

Maybe the internal method `_lincombmul!` could itself directly act on the tuple of maps and apply each of them in a recursive manner?

Reply:
Yeah, I think I got something working, and it is, as usual with recursion, simply beautiful... I believe. The distinction between the ThreeArg route and the FiveArg route is now crystal clear: it's whether we allocate an intermediate `z = similar(y)` or not. Everything else is just the same.